sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -203,7 +203,11 @@ object Cast {
       > SELECT _FUNC_('10' as int);
        10
   """)
-case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String] = None)
+case class Cast(
+    child: Expression,
+    dataType: DataType,
+    timeZoneId: Option[String] = None,
+    timestampScaleFactor: Long = MICROS_PER_SECOND)
```

Comment:
`Cast` is a widely used expression. Maybe a cleaner solution is to create a new expression `GetMicroSeconds` to convert a timestamp to microseconds count.

Reply:
I have to preserve semantic of `Cast` in `ParseToTimestamp`. I could create a parent class for `Cast` like `BaseCast` with `timestampScaleFactor`, and inherit it in `Cast` with `timestampScaleFactor` set to `MICROS_PER_SECOND`. WDYT?

Reply:
> I have to preserve semantic of Cast in ParseToTimestamp.

Can you elaborate on it? For end users, `ParseToTimestamp` and `Cast` are 2 function that are unrelated, though internally `ParseToTimestamp` is implemented via `Cast.`

Reply:
`ParseToTimestamp` extends `RuntimeReplaceable`,  and the optimizer replaces it by `Cast(UnixTimestamp)`. Not sure how it is important.

Frankly speaking, I would add new expression (and register a function) which just takes a string and a pattern, and parse it to `TimestampType` in microsecond precision. We have `DateTimeUtils.stringToTimestamp` but it doesn't accept a pattern.

Surprisingly, parsing of a string with specific pattern is possible by using `from_csv` now which not strait-forward method of doing this.

Reply:
>  For end users, ParseToTimestamp and Cast are 2 functions that are unrelated, though internally ParseToTimestamp is implemented via Cast.

`ParseToTimestamp` has to behave the same as `Cast` when a pattern is not provided.

Reply:
we can do
```
def this(left: Expression, format: Expression) = {
    this(left, Option(format), GetMicroSeconds(UnixTimestamp(left, format))
  }

  def this(left: Expression) = this(left, None, Cast(left, TimestampType))
```